contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/client/ZKClient.java

```diff
@@ -0,0 +1,434 @@
+              promise.fail(KeeperException.create(code, p));
+          }
+        }, null);
+        return promise;
+      }
+    };
+  }
+
+  /**
+   * Returns task to delete znode of the given path if such a node exists
+   * and the given version matches the version of the node (if the given
+   * version is -1, it matches any node's versions).
+   *
+   * @param path path to the znode.
+   * @param version expected matching version.
+   * @return task to delete znode.
+   */
+  public Task<Void> delete(String path, int version) {
+    return Task.async("zkDelete: " + path, () -> {
+      SettablePromise<Void> promise = Promises.settable();
+      _zkClient.delete(path, version, (int rc, String p, Object ctx) -> {
+        KeeperException.Code code = KeeperException.Code.get(rc);
+        switch (code) {
+          case OK:
+            promise.done(null);
+            break;
+          default:
+            promise.fail(KeeperException.create(code, p));
+        }
+      }, null);
+      return promise;
+    });
+  }
+
+  /**
+   * Returns task that will execute all the given {@link Op operation}s in
+   * an atomic manner.
+   *
+   * @param ops operations to execute.
+   * @param executor {@code Executor} that will be used to run the operations.
+   * @return task to execute multiple operations.
+   */
+  public Task<List<OpResult>> multi(Iterable<Op> ops, Executor executor) {
+    return Task.blocking(() -> _zkClient.multi(ops), executor);
+  }
+
+  /**
+   * Returns task that will wait for the given {@link KeeperState} to fulfil.
+   * The task will be failed if the underlying zookeeper session expires.
+   *
+   * @param state keeper state to wait for.
+   * @return task that waits for a certain keeper state.
+   */
+  public Task<Void> waitFor(KeeperState state) {
+    return Task.async("waitFor " + state.toString(), () -> {
+      SettablePromise<Void> promise = Promises.settable();
+      synchronized (_mutex) {
+        if (_currentState == state) {
+          return Promises.VOID;
+        } else {
+          _listener.subscribe(state, (p) -> {
+            if (p.isFailed()) {
+              promise.fail(p.getError());
+            } else {
+              promise.done(null);
+            }
+          });
+        }
+      }
+      return promise;
+    });
+  }
+
+  public Task<Void> waitFor(KeeperState state, long deadline) {
+    return waitFor(state).withTimeout(deadline - System.currentTimeMillis(), TimeUnit.MILLISECONDS);
+  }
+
+  public Task<String> ensurePathExists(String path) {
+    return exists(path).flatMap(stat -> {
+      if (!stat.isPresent()) {
+        String parent = path.substring(0, path.lastIndexOf('/'));
+        if (parent.isEmpty()) { // parent is root
+          return create(path, null, ZooDefs.Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT);
+        } else {
+          return ensurePathExists(parent)
+              .flatMap(unused -> create(path, null, ZooDefs.Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT));
```

Comment:
The javadoc of andThen(Task t) says that "Task passed in as a parameter will run even if the previous task completes fails", which is not what we want. Here we don't want to run the create task if ensurePathExists fails.